### PR TITLE
 Add settings() method

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This adds a Firestore.settings() method similar to what the Web SDK has. The settings provided in this method will be merged with the settings provided at construction time. The settings() method can only be called once before any other operations.

With this PR, it will be possible to enable `timestampsInSnapshots` without any further API changes.

### CODE REVIEW

**Comment by schmidt-sebastian** [comment]
Assigning this to @hiranya911 via this very comment.

File: src/index.js [hunk 1]
```diff
@@ -250,7 +250,15 @@ class Firestore {
      * @private
      * @type {Object}
      */
-    this._initalizationOptions = options;
+    this._initalizationOptions = settings;
```
Comment (hiranya911) [inline src/index.js]: Should this be renamed to _initializationSettings for consistency?

File: src/index.js [hunk 2]
```diff
@@ -313,6 +296,43 @@ follow these steps, YOUR APP MAY BREAK.`);
     Firestore.log('Firestore', null, 'Initialized Firestore');
   }
 
+  /**
+   * Specifies custom settings to be used to configure the `Firestore`
+   * instance. Can only be invoked once and before any other Firestore method.
+   *
+   * If settings are provided via both `settings()` and the Firestore
```
Comment (hiranya911) [inline src/index.js]: I think you can code format `Firestore` here.

File: src/index.js [hunk 3]
```diff
@@ -313,6 +296,43 @@ follow these steps, YOUR APP MAY BREAK.`);
...
+   *
+   * @param {object} settings The settings to use for all Firestore operations.
+   */
+  settings(settings) {
+    validate.isObject('settings', settings);
+    validate.isOptionalString('settings.projectId', settings.projectId);
+    validate.isOptionalBoolean(
+        'settings.timestampsInSnapshots', settings.timestampsInSnapshots);
+
+    if (this._clientInitialized) {
```
Comment (hiranya911) [inline src/index.js]: May be start with these 2 checks, before validating the `settings` argument.

File: test/index.js [hunk 4]
```diff
@@ -314,6 +314,48 @@ describe('instantiation', function() {
...
+
+  it('cannot change settings after client initialized', function() {
+    return createInstance().then(firestore => {
+      assert.throws(
+          () => firestore.settings({}),
+          /Firestore has already been started and its settings can no longer be changed. You can only call settings\(\) before calling any other methods on a Firestore object./);
+    });
+  });
+
+  it('validates project ID is string', function() {
```
Comment (hiranya911) [inline test/index.js]: Add similar test for validating the value of `timestampsInSnapshots`.

File: src/index.js (hunk 1 above)
Comment (schmidt-sebastian) [reply src/index.js]: That makes sense. Done.

File: src/index.js (hunk 2 above)
Comment (schmidt-sebastian) [reply src/index.js]: Done (here and in the `.d.ts` file).

File: src/index.js (hunk 3 above)
Comment (schmidt-sebastian) [reply src/index.js]: Most other functions call these validation helpers right away, so I am inclined to leave it as is. Moving these helper calls further down also looks a little "goofy" to me.

File: test/index.js (hunk 4 above)
Comment (schmidt-sebastian) [reply test/index.js]: Done (also added the missing validation in the constructor).

**Comment by schmidt-sebastian** [comment]
Adding @mikelehen for OWNERS.

**Pull request merged**